oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -135,11 +153,26 @@ public AccessToken refreshAccessToken() throws IOException {
 
   @Override
   public String retrieveSubjectToken() throws IOException {
+    // AWS IDMSv2 introduced a requirement for a session token to be present
+    // with the requests made to metadata endpoints. This requirement is to help
+    // prevent SSRF attacks.
+    // Presence of "imdsv2_session_token_url" in Credential Source of config file
+    // will trigger a flow with session token, else there will not be a session
+    // token with the metadata requests.
+    // Both flows work for IDMS v1 and v2. But if IDMSv2 is enabled, then if
+    // session token is not present, Unauthorized exception will be thrown.
+    Map<String, Object> metadataRequestHeaders = new HashMap<>();
+    if (awsCredentialSource.imdsv2SessionTokenUrl != null) {
```

Comment:
i think this url check has to be inside the method. Otherwise the method assumes the check happens outside, but that can change ) The getAwsImdsv2SessionToken can return either null or token. And this is what you can check here before adding the result into the map.

Reply:
Refactored to a method that prepares the headers.